### PR TITLE
Fix ONNX export for causal LM sequence classifiers by removing reverse indexing

### PR DESCRIPTION
**Pull request by dwyatte**
# What does this PR do?

Follow-up to https://github.com/huggingface/transformers/pull/27450 and another step to fixing https://github.com/huggingface/optimum/issues/1527. ONNX implements indexing using a combination of its own operators and when using reverse indexing (e.g., -1 to indicate 1 element from the right side of an array), it can produce incorrect results (see [PyTorch's ONNX export code](https://github.com/pytorch/pytorch/blob/71bedc3a69e3203fd8f76a68ecf2bd7c58d2e13e/torch/onnx/symbolic_opset9.py#L5859-L5865)). In practice, this can cause the batch dimension to get shuffled

Causal LM sequence were previously using `-1` for the last token. Adding `sequence_lengths = torch.where(sequence_lengths >= 0, sequence_lengths, input_ids.shape[-1] - 1)` effectively removes reverse indexing

While this could be fixed in https://github.com/huggingface/optimum by forcing the inputs used to trace the graph to contain a pad token and avoiding reverse indexing, it seems better to fix in `transformers` with the added benefit of bringing the code in line with TensorFlow implementations of the same code (e.g., https://github.com/huggingface/transformers/pull/25085/files#diff-7c6fdd54ac4b8ce0c09bb17da15f176d3e5827df39dd8234fd802631e99ef38dR801-R804)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker, @amyeroberts, @younesbelkada (CC @fxmarty)

### CODE REVIEW

File: src/transformers/models/ctrl/modeling_ctrl.py
Comment (amyeroberts) [inline src/transformers/models/ctrl/modeling_ctrl.py:803]: * Let's split across lines - it'll use fewer lines overall
* Can we instead use modulo to convert to the equivalent negative index? The logic at the moment assumes we only ever want to take the final index and I think it'll be faster than torch.where 

```suggestion
                sequence_lengths = sequence_lengths % input_ids.shape[-1]
                sequence_lengths = sequence_lengths.to(logits.device)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing this! 

General comment on the indexing assumptions made here

File: src/transformers/models/ctrl/modeling_ctrl.py
Comment (dwyatte) [reply src/transformers/models/ctrl/modeling_ctrl.py]: That should work. Updated all changes

**Review by amyeroberts — APPROVED** [review]
LGTM - thanks for iterating! 

I'd like to have another approval from either @younesbelkada or @ArthurZucker before merging as they know the causal lm models well

**Review by ArthurZucker — APPROVED** [review]
Thanks both, looks good yep! 
More models that can benefit from this: and that use `sequence_lengths = (torch.ne(input_ids, self.config.pad_token_id).sum(dim=-1) - 1).to(logits.device)` 
(not sure if argmax is faster?) 

- bloom 
- falcon
- mpt

let's try to unify this!

**Comment by dwyatte** [comment]
Thanks @ArthurZucker @amyeroberts. I've unified the Bloom/Falcon/MPT implementations, but doing so triggered what looks to be an unrelated CI failure. Can someone take a look and fix/disable that test if it is indeed unrelated?

```
FAILED tests/models/seamless_m4t/test_modeling_seamless_m4t.py::SeamlessM4TModelWithTextInputTest::test_retain_grad_hidden_states_attentions - AttributeError: 'NoneType' object has no attribute 'retain_grad'
```

**Comment by amyeroberts** [comment]
@dwyatte Yep, that's a flaky test. A patch to skip it in the testing suite was recently merged into main to prevent it affecting unrelated PRs like this one :) Could you rebase to include recent updates and trigger a new CI run?

**Comment by dwyatte** [comment]
@amyeroberts Hm, [b0db02c](https://github.com/huggingface/transformers/commit/b0db02c395cee1e8b2ea73077c798617d14be2b4) is your latest commit on main, so I think `tests/models/seamless_m4t/test_modeling_seamless_m4t.py::SeamlessM4TModelWithTextInputTest::test_retain_grad_hidden_states_attentions` is still broken/flaking there

**Comment by amyeroberts** [comment]
@dwyatte hm, that's odd. The test shouldn't even be running as [it's explicitly skipped](https://github.com/huggingface/transformers/blob/224ab70969d1ac6c549f0beb3a8a71e2222e50f7/tests/models/seamless_m4t/test_modeling_seamless_m4t.py#L616). In your local, on this branch, do you see this skip condition in `test_modeling_seamless_m4t.py`?

**Comment by dwyatte** [comment]
@amyeroberts I see what's going on -- the failure is on `SeamlessM4TModelWithTextInputTest` but the explicit skip exists on `SeamlessM4TModelWithSpeechInputTest`. Let me know if I should add the same skip to `SeamlessM4TModelWithTextInputTest` on my branch or if you prefer a different fix/PR

**Comment by dwyatte** [comment]
Ok @amyeroberts @ArthurZucker, after rebasing on the above, this is ready for merging. Thanks both!

**Comment by sentialx** [comment]
Why not just have a shared util for this, instead of repeating the code all over the place

**Pull request merged**
